src/school/teachers/dqn_teacher.py

```diff
@@ -0,0 +1,223 @@
+from .teacher import Teacher
+from typing import List
+from ..task import Task
+from .models.dqn import *
+from .losses import dqn_loss
+import tensorflow as tf
+from random import shuffle
+from .. import Result
+from copy import deepcopy
+from .utils.dqn_structs import *
+from random import random, randint
+
+
+class DQNTeacher(Teacher):
+    """Deep Q-learning agent."""
+
+    def __init__(self, nSkills: int, tasks: List[Task], timeToExam=None, noExamTasks=None,
+                 **kwargs):
+        """Set parameters, initialize network."""
+        super().__init__(nSkills, tasks, **kwargs)
+        self.memSize = 5000
+        self.noTargetIte = 2000
+        self.noLearnIte = 50
+        self.batchSize = 500
+        self.epsilon = 0.4
+        self.minEpsilon = 0.005
+        self.epsilonDecreaseFactor = 1  # value between 0 and 1
+        self.lossFun = dqn_loss
+        modelInputSize = len(tasks) * 2
+        self.estimator = DQN(modelInputSize)
+        self.targetEstimator = DQN(modelInputSize)
+        self.studentsStates = {}
+        self.studentsPrevState = {}
+        self.beforeExamTasks = {}
+        self.learnCounter = 0
+        self.gamma = 0.8
+        self.mem = [MemoryRecord() for _ in range(self.memSize)]
+        self.memIdx = -1
+        self.batchSelectIdxs = list(range(self.memSize))
+        self.timeToExam = timeToExam
+        self.noTasks = len(tasks)
+        self.noExamTasks = noExamTasks
+        self.memFull = False
+        self.epsilonDecreaseMultiplayer = (self.epsilonDecreaseFactor / self.nSkills) * 1e-5
+        self.__examResults = {}
+        self.__statesBuff = [0] * self.batchSize
+        self.__actionsBuff = [0] * self.batchSize
+        self.__realQsBuff = [None] * self.batchSize
+        self.__targetCounter = 0
+        self.__innerStudentsStates = {}
+
+    def choose_task(self, student) -> Task:
+        if student.id not in self.studentsStates:
+            self.studentsStates[student.id] = self.__get_new_student_state()
+        # exploitation
+        if random() > self.epsilon:
+            studentState = tf.expand_dims(tf.constant(self.studentsStates[student.id]), axis=0)
+            actionsValues = self.estimator(studentState)[0]
+            chosenTaskId = tf.argmax(actionsValues).numpy()
+        # exploration
+        else:
+            chosenTaskId = randint(0, self.noTasks - 1)
+        return [task_ for task_ in self.tasks if task_.id == chosenTaskId][0]
+
+    def receive_result(self, result, last=False, reward=None) -> None:
+        if result.task is None:  # Added cause my stupid changes in Classroom
+            return
```

Comment:
Wrzuciłem to, bo w Classroomie podajemy taki Result, wymagamy go w AC, a widziałem, że wyrzuciliście tamtą linijkę, więc ją przywróciłem i dałem zabezpieczenie

Reply:
Ok

Reply:
No to apropo tego, mam prośbę żebyś to zmienił w classroomie, żeby ten receive result był tylko uruchamiany dla Twojego techeara bo tak to ja będę musiał to też dodać takie zabezpieczenie do base_teachera i w przyszłości do każdego innego teachera.